synapse/storage/_base.py

```diff
@@ -275,6 +287,56 @@ def _check_safe_to_upsert(self):
                 self._check_safe_to_upsert,
             )
 
+    @defer.inlineCallbacks
+    def _set_expiration_date_when_missing(self):
+        """
+        Retrieves the list of registered users that don't have an expiration date, and
+        adds an expiration date for each of them.
+        """
+
+        def select_users_with_no_expiration_date_txn(txn):
+            """Retrieves the list of registered users with no expiration date from the
+            database.
+            """
+            sql = (
+                "SELECT users.name FROM users"
+                " LEFT JOIN account_validity ON (users.name = account_validity.user_id)"
+                " WHERE account_validity.user_id is NULL;"
+            )
+            txn.execute(sql, [])
+            return self.cursor_to_dict(txn)
+
+        res = yield self.runInteraction(
+            "get_users_with_no_expiration_date",
+            select_users_with_no_expiration_date_txn,
+        )
+
+        if res:
+            for user in res:
+                self.runInteraction(
+                    "set_expiration_date_for_user_background",
+                    self.set_expiration_date_for_user_txn,
+                    user["name"],
+                )
```

Comment:
You want to do the select and insert in the same transaction, otherwise things might change from underneath you.

(If you want to be really really snazzy, you can do this all in one query

```sql
INSERT INTO account_validity SELECT users.name, ?, ? FROM users LEFT JOIN ...
```

)